File: android/app/src/main/java/chat/rocket/android_ddp/DDPClientImpl.java
package chat.rocket.android_ddp;

import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import bolts.Task;
import bolts.TaskCompletionSource;
import chat.rocket.android_ddp.rx.RxWebSocket;
import chat.rocket.android_ddp.rx.RxWebSocketCallback;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import okhttp3.OkHttpClient;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import rx.Observable;
import rx.functions.Func1;
import rx.subscriptions.CompositeSubscription;

public class DDPClientImpl {
  private final DDPClient mClient;
  private final RxWebSocket mWebSocket;
  private Observable<RxWebSocketCallback.Base> mObservable;
  private CompositeSubscription mBaseSubscriptions;

  public DDPClientImpl(DDPClient self, OkHttpClient client) {
    mWebSocket = new RxWebSocket(client);
    mClient = self;
  }

  private static JSONObject toJson(String s) {
    if (TextUtils.isEmpty(s)) return null;
    try {
      return new JSONObject(s);
    } catch (JSONException e) {
      return null;
    }
  }

  private static String extractMsg(JSONObject response) {
    if (response == null || response.isNull("msg")) {
      return null;
    } else {
      return response.optString("msg");
    }
  }

  public void connect(final TaskCompletionSource<DDPClientCallback.Connect> task, final String url,
      String session) {
    try {
      mObservable = mWebSocket.connect(url).autoConnect();
      CompositeSubscription subscriptions = new CompositeSubscription();

      subscriptions.add(mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Open)
          .subscribe(callback -> {
            sendMessage("connect",
                json -> (TextUtils.isEmpty(session) ? json : json.put("session", session)).put(
                    "version", "pre2").put("support", new JSONArray().put("pre2").put("pre1")));
          }, err -> {
          }));

      subscriptions.add(
          mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
              .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
              .map(DDPClientImpl::toJson)
              .subscribe(response -> {
                String msg = extractMsg(response);
                if ("connected".equals(msg) && !response.isNull("session")) {
                  task.setResult(
                      new DDPClientCallback.Connect(mClient, response.optString("session")));
                  subscriptions.unsubscribe();
                } else if ("error".equals(msg) && "Already connected".equals(
                    response.optString("reason"))) {
                  task.setResult(new DDPClientCallback.Connect(mClient, null));
                  subscriptions.unsubscribe();
                } else if ("failed".equals(msg)) {
                  task.setError(
                      new DDPClientCallback.Connect.Failed(mClient, response.optString("version")));
                  subscriptions.unsubscribe();
                }
              }, err -> {
              }));

      addErrorCallback(subscriptions, task);

      subscribeBaseListeners();
    } catch (Exception e) {
      Log.e("connect", "Error", e);
    }
  }

  public boolean isConnected() {
    return mWebSocket != null && mWebSocket.isConnected();
  }

  public void ping(final TaskCompletionSource<DDPClientCallback.Ping> task,
      @Nullable final String id) {
    CompositeSubscription subscriptions = new CompositeSubscription();

    subscriptions.add(
        mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
            .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
            .map(DDPClientImpl::toJson)
            .timeout(4, TimeUnit.SECONDS)
            .subscribe(response -> {
              String msg = extractMsg(response);
              if ("pong".equals(msg)) {
                if (response.isNull("id")) {
                  task.setResult(new DDPClientCallback.Ping(mClient, null));
                  subscriptions.unsubscribe();
                } else {
                  String _id = response.optString("id");
                  if (id.equals(_id)) {
                    task.setResult(new DDPClientCallback.Ping(mClient, id));
                    subscriptions.unsubscribe();
                  }
                }
              }
            }, err -> {
              task.setError(new DDPClientCallback.Ping.Timeout(mClient));
            }));

    addErrorCallback(subscriptions, task);

    if (TextUtils.isEmpty(id)) {
      sendMessage("ping", null);
    } else {
      sendMessage("ping", json -> json.put("id", id));
    }
  }

  public void sub(final TaskCompletionSource<DDPSubscription.Ready> task, String name,
      JSONArray params, String id) {
    CompositeSubscription subscriptions = new CompositeSubscription();

    subscriptions.add(
        mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
            .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
            .map(DDPClientImpl::toJson)
            .subscribe(response -> {
              String msg = extractMsg(response);
              if ("ready".equals(msg) && !response.isNull("subs")) {
                JSONArray ids = response.optJSONArray("subs");
                for (int i = 0; i < ids.length(); i++) {
                  String _id = ids.optString(i);
                  if (id.equals(_id)) {
                    task.setResult(new DDPSubscription.Ready(mClient, id));
                    subscriptions.unsubscribe();
                    break;
                  }
                }
              } else if ("nosub".equals(msg) && !response.isNull("id") && !response.isNull(
                  "error")) {
                String _id = response.optString("id");
                if (id.equals(_id)) {
                  task.setError(new DDPSubscription.NoSub.Error(mClient, id,
                      response.optJSONObject("error")));
                  subscriptions.unsubscribe();
                }
              }
            }, err -> {
            }));

    addErrorCallback(subscriptions, task);

    sendMessage("sub", json -> json.put("id", id).put("name", name).put("params", params));
  }

  public void unsub(final TaskCompletionSource<DDPSubscription.NoSub> task,
      @Nullable final String id) {
    CompositeSubscription subscriptions = new CompositeSubscription();

    subscriptions.add(
        mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
            .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
            .map(DDPClientImpl::toJson)
            .subscribe(response -> {
              String msg = extractMsg(response);
              if ("nosub".equals(msg) && response.isNull("error") && !response.isNull("id")) {
                String _id = response.optString("id");
                if (id.equals(_id)) {
                  task.setResult(new DDPSubscription.NoSub(mClient, id));
                  subscriptions.unsubscribe();
                }
              }
            }, err -> {
            }));

    addErrorCallback(subscriptions, task);

    sendMessage("unsub", json -> json.put("id", id));
  }

  public void rpc(final TaskCompletionSource<DDPClientCallback.RPC> task, String method,
      JSONArray params, String id, long timeoutMs) {
    CompositeSubscription subscriptions = new CompositeSubscription();

    subscriptions.add(
        mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
            .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
            .map(DDPClientImpl::toJson)
            .timeout(timeoutMs, TimeUnit.MILLISECONDS)
            .subscribe(response -> {
              String msg = extractMsg(response);
              if ("result".equals(msg)) {
                String _id = response.optString("id");
                if (id.equals(_id)) {
                  if (!response.isNull("error")) {
                    task.setError(new DDPClientCallback.RPC.Error(mClient, id,
                        response.optJSONObject("error")));
                  } else {
                    String result = response.optString("result");
                    task.setResult(new DDPClientCallback.RPC(mClient, id, result));
                  }
                  subscriptions.unsubscribe();
                }
              }
            }, err -> {
              if (err instanceof TimeoutException) {
                task.setError(new DDPClientCallback.RPC.Timeout(mClient));
              }
            }));

    addErrorCallback(subscriptions, task);

    sendMessage("method", json -> json.put("method", method).put("params", params).put("id", id));
  }

  private void subscribeBaseListeners() {
    if (mBaseSubscriptions != null &&
        mBaseSubscriptions.hasSubscriptions() && !mBaseSubscriptions.isUnsubscribed()) {
      return;
    }

    mBaseSubscriptions = new CompositeSubscription();
    mBaseSubscriptions.add(
        mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
            .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
            .map(DDPClientImpl::toJson)
            .subscribe(response -> {
              String msg = extractMsg(response);
              if ("ping".equals(msg)) {
                if (response.isNull("id")) {
                  sendMessage("pong", null);
                } else {
                  sendMessage("pong", json -> json.put("id", response.getString("id")));
                }
              }
            }, err -> {
            }));
  }

  public Observable<DDPSubscription.Event> getDDPSubscription() {
    String[] targetMsgs = { "added", "changed", "removed", "addedBefore", "movedBefore" };
    return mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Message)
        .map(callback -> ((RxWebSocketCallback.Message) callback).responseBodyString)
        .map(DDPClientImpl::toJson)
        .filter(response -> {
          String msg = extractMsg(response);
          for (String m : targetMsgs) {
            if (m.equals(msg)) return true;
          }
          return false;
        })
        .map((Func1<JSONObject, DDPSubscription.Event>) response -> {
          String msg = extractMsg(response);
          if ("added".equals(msg)) {
            return new DDPSubscription.Added(mClient, response.optString("collection"),
                response.optString("id"),
                response.isNull("fields") ? null : response.optJSONObject("fields"));
          } else if ("addedBefore".equals(msg)) {
            return new DDPSubscription.Added.Before(mClient, response.optString("collection"),
                response.optString("id"),
                response.isNull("fields") ? null : response.optJSONObject("fields"),
                response.isNull("before") ? null : response.optString("before"));
          } else if ("changed".equals(msg)) {
            return new DDPSubscription.Changed(mClient, response.optString("collection"),
                response.optString("id"),
                response.isNull("fields") ? null : response.optJSONObject("fields"),
                response.isNull("cleared") ? new JSONArray() : response.optJSONArray("before"));
          } else if ("removed".equals(msg)) {
            return new DDPSubscription.Removed(mClient, response.optString("collection"),
                response.optString("id"));
          } else if ("movedBefore".equals(msg)) {
            return new DDPSubscription.MovedBefore(mClient, response.optString("collection"),
                response.optString("id"),
                response.isNull("before") ? null : response.optString("before"));
          }

          return null;
        })
        .asObservable();
  }

  public void unsubscribeBaseListeners() {
    if (mBaseSubscriptions.hasSubscriptions() && !mBaseSubscriptions.isUnsubscribed()) {
      mBaseSubscriptions.unsubscribe();
    }
  }

  public Task<RxWebSocketCallback.Close> getOnCloseCallback() {
    TaskCompletionSource<RxWebSocketCallback.Close> task = new TaskCompletionSource<>();

    mObservable.filter(callback -> callback instanceof RxWebSocketCallback.Close)
        .cast(RxWebSocketCallback.Close.class)
        .subscribe(close -> {
          task.setResult(close);
        }, err -> {
          if (err instanceof Exception) {
            task.setError((Exception) err);
          } else {
            task.setError(new Exception(err));
          }
        });

    return task.getTask().onSuccessTask(_task -> {
      unsubscribeBaseListeners();
      return _task;
    });
  }

  private void sendMessage(String msg, @Nullable JSONBuilder json) {
    try {
      JSONObject origJson = new JSONObject().put("msg", msg);
      String msg2 = (json == null ? origJson : json.create(origJson)).toString();
      mWebSocket.sendText(msg2);
    } catch (Exception e) {
      Log.e("ANDROID_DDP", "sendMessage", e);
    }
  }

  private void addErrorCallback(CompositeSubscription subscriptions, TaskCompletionSource<?> task) {
    subscriptions.add(mObservable.subscribe(base -> {
    }, err -> {
      task.setError(new Exception(err));
      subscriptions.unsubscribe();
    }));
  }

  public void close(int code, String reason) {
    try {
      mWebSocket.close(code, reason);
    } catch (Exception e) {
      Log.e("ANDROID_DDP", "addErrorCallback", e);
    }
  }

  private interface JSONBuilder {
    @NonNull
    JSONObject create(JSONObject root) throws JSONException;
  }
}

File: android/app/src/main/java/chat/rocket/reactnative/voip/ddp/DdpClientImpl.kt
package chat.rocket.reactnative.voip.ddp

import android.os.Handler
import android.os.Looper
import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class DdpClientImpl(
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .build()
) : DdpClient {

    private data class QueuedMethodCall(
        val method: String,
        val params: JSONArray,
        val callback: (Boolean) -> Unit
    )

    companion object {
        private const val TAG = "RocketChat.DdpClient"
    }

    private var webSocket: WebSocket? = null
    private var sendCounter = 0
    private var isConnected = false
    private val mainHandler = Handler(Looper.getMainLooper())

    private val pendingCallbacks = mutableMapOf<String, (JSONObject) -> Unit>()
    private val queuedMethodCalls = mutableListOf<QueuedMethodCall>()
    private var connectedCallback: ((Boolean) -> Unit)? = null

    override var onCollectionMessage: ((JSONObject) -> Unit)? = null

    override fun connect(host: String, callback: (Boolean) -> Unit) {
        val wsUrl = buildWebSocketURL(host)

        Log.d(TAG, "Connecting to $wsUrl")

        val request = Request.Builder().url(wsUrl).build()

        webSocket = httpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket opened")
                val connectMsg = JSONObject().apply {
                    put("msg", "connect")
                    put("version", "1")
                    put("support", JSONArray().apply {
                        put("1"); put("pre2"); put("pre1")
                    })
                }
                webSocket.send(connectMsg.toString())
                waitForConnected(10_000L, callback)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket failure: ${t.message}")
                isConnected = false
                mainHandler.post { callback(false) }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed: $code $reason")
                isConnected = false
            }
        })
    }

    override fun login(token: String, callback: (Boolean) -> Unit) {
        val msg = nextMessage("method").apply {
            put("method", "login")
            put("params", JSONArray().apply {
                put(JSONObject().apply { put("resume", token) })
            })
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val hasError = data.has("error")
                if (hasError) {
                    Log.e(TAG, "Login failed: ${data.opt("error")}")
                } else {
                    Log.d(TAG, "Login succeeded")
                }
                mainHandler.post { callback(!hasError) }
            }
        }

        if (!send(msg)) {
            mainHandler.post { callback(false) }
        }
    }

    override fun subscribe(name: String, params: JSONArray, callback: (Boolean) -> Unit) {
        val msg = nextMessage("sub").apply {
            put("name", name)
            put("params", params)
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val didSubscribe = data.optString("msg") == "ready" && !data.has("error")
                if (didSubscribe) {
                    Log.d(TAG, "Subscribed to $name")
                } else {
                    Log.e(TAG, "Failed to subscribe to $name: ${data.opt("error") ?: "nosub"}")
                }
                mainHandler.post { callback(didSubscribe) }
            }
        }

        if (!send(msg)) {
            synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
            mainHandler.post { callback(false) }
        }
    }

    override fun disconnect() {
        Log.d(TAG, "Disconnecting")
        isConnected = false
        synchronized(pendingCallbacks) { pendingCallbacks.clear() }
        clearQueuedMethodCalls()
        connectedCallback = null
        onCollectionMessage = null
        webSocket?.close(1000, null)
        webSocket = null
        httpClient.dispatcher.executorService.shutdown()
    }

    private fun nextMessage(msg: String): JSONObject {
        sendCounter++
        return JSONObject().apply {
            put("msg", msg)
            put("id", "ddp-$sendCounter")
        }
    }

    private fun send(json: JSONObject): Boolean {
        val ws = webSocket ?: return false
        return ws.send(json.toString())
    }

    override fun callMethod(method: String, params: JSONArray, callback: (Boolean) -> Unit) {
        val msg = nextMessage("method").apply {
            put("method", method)
            put("params", params)
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val hasError = data.has("error")
                if (hasError) {
                    Log.e(TAG, "Method $method failed: ${data.opt("error")}")
                }
                mainHandler.post { callback(!hasError) }
            }
        }

        if (!send(msg)) {
            synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
            mainHandler.post { callback(false) }
        }
    }

    override fun queueMethodCall(method: String, params: JSONArray, callback: (Boolean) -> Unit) {
        synchronized(queuedMethodCalls) {
            queuedMethodCalls.add(
                QueuedMethodCall(
                    method = method,
                    params = params,
                    callback = callback
                )
            )
        }
    }

    override fun hasQueuedMethodCalls(): Boolean =
        synchronized(queuedMethodCalls) { queuedMethodCalls.isNotEmpty() }

    override fun flushQueuedMethodCalls() {
        val queuedCalls = synchronized(queuedMethodCalls) {
            queuedMethodCalls.toList().also { queuedMethodCalls.clear() }
        }

        queuedCalls.forEach { queuedCall ->
            callMethod(queuedCall.method, queuedCall.params, queuedCall.callback)
        }
    }

    override fun clearQueuedMethodCalls() {
        synchronized(queuedMethodCalls) {
            queuedMethodCalls.clear()
        }
    }

    private fun waitForConnected(timeoutMs: Long, callback: (Boolean) -> Unit) {
        connectedCallback = callback
        mainHandler.postDelayed({
            val cb = connectedCallback ?: return@postDelayed
            connectedCallback = null
            Log.e(TAG, "Connect timeout")
            cb(false)
        }, timeoutMs)
    }

    private fun handleMessage(text: String) {
        val json = try {
            JSONObject(text)
        } catch (e: Exception) {
            return
        }

        when (json.optString("msg")) {
            "connected" -> {
                isConnected = true
                mainHandler.removeCallbacksAndMessages(null)
                val cb = connectedCallback
                connectedCallback = null
                cb?.let { mainHandler.post { it(true) } }
            }

            "ping" -> {
                send(JSONObject().apply { put("msg", "pong") })
            }

            "result" -> {
                val id = json.optString("id")
                val cb = synchronized(pendingCallbacks) { pendingCallbacks[id] }
                cb?.invoke(json)
            }

            "ready" -> {
                val subs = json.optJSONArray("subs")
                if (subs != null) {
                    for (index in 0 until subs.length()) {
                        val subId = subs.optString(index)
                        if (subId.isEmpty()) {
                            continue
                        }

                        val cb = synchronized(pendingCallbacks) { pendingCallbacks[subId] }
                        cb?.invoke(json)
                    }
                }
            }

            "changed", "added", "removed" -> {
                onCollectionMessage?.invoke(json)
            }

            "nosub" -> {
                val id = json.optString("id")
                val cb = synchronized(pendingCallbacks) { pendingCallbacks[id] }
                cb?.invoke(json)
            }

            else -> {
                if (json.has("collection")) {
                    onCollectionMessage?.invoke(json)
                }
            }
        }
    }

    private fun buildWebSocketURL(host: String): String {
        var normalizedHost = host.trimEnd('/')

        val useSsl: Boolean
        when {
            normalizedHost.startsWith("https://") -> {
                useSsl = true
                normalizedHost = normalizedHost.removePrefix("https://")
            }
            normalizedHost.startsWith("http://") -> {
                useSsl = false
                normalizedHost = normalizedHost.removePrefix("http://")
            }
            else -> {
                useSsl = true
            }
        }

        val scheme = if (useSsl) "wss" else "ws"
        return "$scheme://$normalizedHost/websocket"
    }
}
